### PR TITLE
Align the stack pointer at a 16-byte boundary at function calls

### LINKED ISSUE #8: Stack is not aligned to a 16-byte boundary before executing the ELF binary
### Problem

Current builds of glibc use SSE instructions to read/write data from/to stack, especially when dealing with variadic arguments (eg, `printf`).

According to the x86_64 ABI, the stack must be aligned to a 16-byte boundary when a function is entered and gcc generates code that (rightly) assumes so.

### Repro

Compile the following program as a dynamically-linked ELF and run it with a different number of arguments:

```c
// gcc -o helloworld helloworld.c

#include <stdio.h>

int main(int argc, char **argv)
{
    for (int i = 0; i < argc; ++i)
    {
        printf("[%d] %s\n", i, argv[i]);
    }
    return 0;
}
```

- Example of 'good' invokation:
`qemu-guest -k app-elfloader_kvm-x86_64 -i rootfs/ld-linux-x86-64.so.2 -e rootfs -a "/ld-linux-x86-64.so.2 --library-path / /helloworld dummy1"`

- Example of 'bad invokation' (odd number of arguments pushed to the stack --> crash):
`qemu-guest -k app-elfloader_kvm-x86_64 -i rootfs/ld-linux-x86-64.so.2 -e rootfs -a "/ld-linux-x86-64.so.2 --library-path / /helloworld dummy1 dummy2"`

I have also attached a zip archive with the exact [binaries](https://github.com/unikraft/app-elfloader/files/11009004/rootfs.zip) I am using.

### Environment

gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
ld.so (Ubuntu GLIBC 2.35-0ubuntu3.1) stable release version 2.35

### PR DESCRIPTION
**Pull request by csvancea**
x86_64 callees assume that the stack is 16-byte aligned. Ignoring this requirement might lead to crashes if the Linux ELFs use vectorized instructions to read from/write to the stack (eg: for copying arguments).

This PR fixes the problem by counting how many arguments are supposed to be pushed to the stack and if the number is odd, we push one more argument on top of the stack, before the auxiliary vector. This guarantees that the stack will be properly aligned when the execution starts at the ELF entry point.

Closes #8.

### CODE REVIEW

**Review by skuenzer — CHANGES_REQUESTED** [review]
Hi @csvancea,

thanks a lot for this PR. Can you update it so that we can achieve better architecture independence by using `UKARCH_SP_ALIGN`? I assume we then need to take into account the number of auxiliary vector entries that we push on the stack, as well. The auxiliary vector is like the environment variables and the application arguments: It is just a list of entries of size: `2 * sizeof(long)`: https://refspecs.linuxfoundation.org/LSB_1.3.0/IA64/spec/auxiliaryvector.html It does not have to be fixed size.

Thanks,

Simon

File: elf_ctx.c [hunk 1]
```diff
@@ -123,41 +124,60 @@ void elf_ctx_init(struct ukarch_ctx *ctx, struct elf_prog *prog,
 		for (char **env = environ; *env; ++env)
 			++envc;
 
+	/* list of all auxiliary vector entries */
+	struct aux_vector_entry auxv[] = {
+		{ AT_NULL, 0x0 },
```
Comment (skuenzer) [inline elf_ctx.c]: For readability, maybe put these entries in the right order and iterate in reverse order over them when pushing to the stack.

Comment (skuenzer) [reply elf_ctx.c]: Similar to `environ` and `argv`, I would push the terminating `AT_NULL` before entering the "stackpush-loop" (and not part of this array). This way the code could potentially cleanly handle an empty `auxv` vector.

File: elf_ctx.c [hunk 2]
```diff
@@ -123,41 +124,60 @@ void elf_ctx_init(struct ukarch_ctx *ctx, struct elf_prog *prog,
...
+	 * calls. (eg: x86_64 requires that the stack is 16-byte aligned)
+	 *
+	 * Here we count how many values are pushed to the stack:
+	 * - auxiliary vector (2 values for each entry);
+	 * - environment variables (plus extra NULL pointer);
+	 * - arguments (plus extra NULL pointer and argument count)
 	 */
+	stackvalc = AUXV_VALUES_PER_ENTRY * ARRAY_SIZE(auxv) + (envc+1) + ((argc-1)+2);
+	stackval_len = stackvalc * sizeof(void *);
+	ctx->sp -= (UKARCH_SP_ALIGN - (stackval_len & UKARCH_SP_ALIGN_MASK)) & UKARCH_SP_ALIGN_MASK;
```
Comment (skuenzer) [inline elf_ctx.c]: Can you use the `ALIGN_DOWN()` or `ALIGN_UP()` macro instead (`<uk/essentials.h>`)? It does basically the same but is maybe quicker to read. Ideally, the computation does not do any assumption on the current alignemnt of the stack and just aligns it properly. We can then later add more things on the stack before the padding without risking to break any of the alignment logic (e.g., a copy of `argv` and `envp` strings: http://articles.manugarg.com/aboutelfauxiliaryvectors.html)

File: elf_ctx.c [hunk 3]
```diff
@@ -123,41 +124,60 @@ void elf_ctx_init(struct ukarch_ctx *ctx, struct elf_prog *prog,
...
-	 * will be read by the application's C library (i.e. argc in the end)
+	 * We need to respect the stack alignment ABI requirements at function
+	 * calls. (eg: x86_64 requires that the stack is 16-byte aligned)
+	 *
+	 * Here we count how many values are pushed to the stack:
+	 * - auxiliary vector (2 values for each entry);
+	 * - environment variables (plus extra NULL pointer);
+	 * - arguments (plus extra NULL pointer and argument count)
 	 */
+	stackvalc = AUXV_VALUES_PER_ENTRY * ARRAY_SIZE(auxv) + (envc+1) + ((argc-1)+2);
```
Comment (skuenzer) [inline elf_ctx.c]: Instead of AUXV_VALUES_PER_ENTRY, you could also use `sizeof(struct aux_vector_entry)`.

File: elf_ctx.c [hunk 4]
```diff
@@ -123,41 +124,60 @@ void elf_ctx_init(struct ukarch_ctx *ctx, struct elf_prog *prog,
...
-	push_auxv(ctx, AT_SYSINFO, 0x0);
-	push_auxv(ctx, AT_EXECFN, 0x0);
-	push_auxv(ctx, AT_DCACHEBSIZE, 0x0);
-	push_auxv(ctx, AT_ICACHEBSIZE, 0x0);
-	push_auxv(ctx, AT_UCACHEBSIZE, 0x0);
-	push_auxv(ctx, AT_NOTELF, 0x0);
-	push_auxv(ctx, AT_PLATFORM, (uintptr_t) auxv_platform);
+	for (i = 0; i < (int) ARRAY_SIZE(auxv); ++i) {
+		ukarch_rctx_stackpush(ctx, auxv[i].val);
+		ukarch_rctx_stackpush(ctx, auxv[i].key);
```
Comment (skuenzer) [inline elf_ctx.c]: Does `ukarch_rctx_stackpush(ctx, auxv[i]);` work, too? If so, I'd prefer that as good practice: Preferably deriving everything from the `struct auxv_entry` definition.

File: elf_ctx.c [hunk 5]
```diff
@@ -83,23 +83,24 @@
...
-		ukarch_rctx_stackpush(ctx, (long) val);		\
-		ukarch_rctx_stackpush(ctx, (long) key);		\
-	} while (0)
-
 void elf_ctx_init(struct ukarch_ctx *ctx, struct elf_prog *prog,
 		  int argc, char *argv[], char *environ[],
 		  uint64_t rand[2])
 {
-	int i, envc;
+	int i, envc, stackvalc, stackval_len;
```
Comment (skuenzer) [inline elf_ctx.c]: Maybe you want to cal it `elfvec_len` for `ELF vectors length`? The count may no be needed if you update the computation formular.

File: elf_ctx.c [hunk 6]
```diff
@@ -83,23 +83,24 @@
 #define AT_SYSINFO_EHDR		33
 #define AT_SYSINFO		32
 
+#define AUXV_VALUES_PER_ENTRY 	2
+
+struct aux_vector_entry {
```
Comment (skuenzer) [inline elf_ctx.c]: I'd probably call it `auxv_entry`

File: elf_ctx.c (hunk 2 above)
Comment (skuenzer) [reply elf_ctx.c]: I am thinking of something like this to do the padding (please double-check):
```c
ctx->sp = ALIGN_DOWN(ctx->sp - elfvec_len, UKARCH_SP_ALIGN) + elfvec_len;
```

**Review by skuenzer — CHANGES_REQUESTED** [review]
Hi @csvancea, thanks a lot for this revision. Since your second commit removes all changes from the first commit it is probably not worth to keep the first commit. You can merge them together so that we have only one commit for this PR. I have some minor suggestions to improve code readability but other from this, it looks fine. 👍🏻 
Please also double-check also that `checkpatch` is happy with your code.
Simon

File: elf_ctx.c (hunk 3 above)
Comment (skuenzer) [reply elf_ctx.c]: You could directly compute `elfvec_len` and using `sizeof()`s for each entry item size, like:
```c
elfvec_len = (ARRAY_SIZE(auxv) * sizeof(struct auxv_entry))
             + (envc + 1) * sizeof(uintptr_t)
             + (argc + 1) * sizeof(uintptr_t)
             + sizeof(long);
```

File: elf_ctx.c (hunk 2 above)
Comment (csvancea) [reply elf_ctx.c]: Thanks for the improved formula! I always thought that there were ways to simplify it, but couldn't find how to do it.

Compared the result of both calculations and they yield the same results.

File: elf_ctx.c (hunk 4 above)
Comment (csvancea) [reply elf_ctx.c]: It works and it definitely looks better.

File: elf_ctx.c (hunk 1 above)
Comment (csvancea) [reply elf_ctx.c]: I don't think the order matters as long as `AT_NULL` is the first value pushed. Anyway, while reversing the order, I have noticed that `AT_SYSINFO` was being pushed twice, so I have removed one entry.

**Review by skuenzer — APPROVED** [review]
Thanks for the updates!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request closed without merging**